skrf/calibration/calibration.py

```diff
@@ -5756,21 +5766,34 @@ def determine_reflect(thru_m, reflect_m, line_m, reflect_approx=None,
     # tt is equal to T from equation (24) in the paper
     tt = einsum('ijk,ikl -> ijl', rd, inv(rt))
 
+    # regularize the parameters in case of matched thru and line. see gh-870
+    tt[:,1,0] = _regularize_inplace(tt[:,1,0])
+    tt[:,0,1] = _regularize_inplace(tt[:,0,1])
+
     a = tt[:,1,0]
     b = tt[:,1,1]-tt[:,0,0]
     c = -tt[:,0,1]
+    sqrtD = sqrt(b*b-4*a*c)
 
-    # the sol1 and sol2 correspond to the ratios (r11/r21) and (r12/r22) from
-    # equations (30) and (31) in the paper
-    sol1 = (-b-sqrt(b*b-4*a*c))/(2*a)
-    sol2 = (-b+sqrt(b*b-4*a*c))/(2*a)
+    # The variables a, b, c define a quadratic equation for which the solutions sol1 and sol2 correspond to the
+    # ratios (r11/r21) and (r12/r22) from equations (30) and (31) in the paper
+    # The quadratic equation has solutions sol1 = (-b-sqrt(b*b-4*a*c))/(2*a), sol2 = (-b+sqrt(b*b-4*a*c))/(2*a)
+    # For a=0 these become degenerate. Also the consequtive equations for x1 and x2 contain singularities for a=0 or c=0
+    # We address this by regularizing small values of a and the candidate solutions.
+    # The impact on the final solution is small as for small a one root of the equation goes to infinity,
+    # for the other the series expension in a is -c/b  - a c^2/b^3 + O(a^2)
 
-    e2 = line.s[:,0,1]**2
+    denom = 2*a
+    sol1 = (-b-sqrtD)/denom
+    sol2 = (-b+sqrtD)/denom
 
+    # equation (32)
     x1 = (tt[:,1,0]*sol1 + tt[:,1,1])/(tt[:,0,1]/sol2 + tt[:,0,0])
     x2 = (tt[:,1,0]*sol2 + tt[:,1,1])/(tt[:,0,1]/sol1 + tt[:,0,0])
 
-    rootChoice = [abs(x1[i] - e2[i]) < abs(x2[i] - e2[i]) for i in range(len(x1))]
+    e2 = line.s[:,0,1]**2
+    rootChoice0 = abs(x1 - e2) < abs(x2 - e2) # original choice for the root. see gh-870
+    rootChoice = npy.abs(x1) < npy.abs(x2) # criterium from eq (55)
```

Comment:
This chooses the root based on which one gives lossy line. However it is inaccurate for very low loss line such as waveguide. The original one is also described in the paper and is more robust. Keep the old root choice algorithm.

Reply:
@Ttl I reverted the root choice, but now I have a failing test which I do not know how to fix. The failing test is:
```
        freq = rf.F(434615384.6153846e-9, .7, 1, unit = 'GHz')
        medium = Coaxial.from_attenuation_VF(freq, att = 3.0, VF = .69)
        thru= medium.line(0, 'm')
        line = medium.line(0.12, 'm')
        short = rf.Network(f=freq.f, s=[[[(-1.+0.017870117714376983j)] ] ])

        r = determine_reflect(thru, rf.two_port_reflect(short, short), line)
        npy.testing.assert_array_almost_equal( r.s, short.s)
```
It does pass with the `rootChoice = npy.abs(x1) < npy.abs(x2)` test, but not with the `rootChoice = abs(x1 - e2) < abs(x2 - e2)`. Do you have any suggestions on how to address this? (maybe this is a case that we should not expect to work, but it seems like a reasonable case to me)

Reply:
This passes on current master:

```python
    def test_determine_reflect_regression(self):
        freq = rf.F(434615384.6153846e-9, .7, 1, unit = 'GHz')
        medium = Coaxial.from_attenuation_VF(freq, att = 3.0, VF = .69)
        thru = medium.line(0, 'm')
        line = medium.line(0.12, 'm')
        short = medium.load(-1.+0.017870117714376983j)

        # workaround division by zero
        thru.s[:,0,0] += 1e-7
        thru.s[:,1,1] += 1e-7
        r = determine_reflect(thru, rf.two_port_reflect(short, short), line)
        npy.testing.assert_array_almost_equal(r.s, short.s)
```

Reply:
@ttl Ok. I tried before only to regularize t, but with the regularization on `s` now all tests are passing.